### PR TITLE
✨ Add Support for the Rotation Sensor

### PR DESCRIPTION
**Pull request by WillXuCodes**
Summary:
Setup the C/Cpp API for the Rotation Sensor.

Motivation:
With the rotation sensor being released in the near feature, it wouldn't hurt to have support for it before it comes out for our users to use as soon as they get their hands on it.

Test Plan:
- [x] Tested C/Cpp API with rotation sensor

### CODE REVIEW

File: include/pros/rotation.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+ * 				 The V5 Rotation Sensor port number from 1-21
+ * \return The velocity value or PROS_ERR_F if the operation failed, setting
+ * errno.
+ */
+int32_t rotation_get_velocity(uint8_t port);
+
+/**
+ * Get the Rotation sensor's current rotational position in terms of an angle.
+ *
+ * TODO: Add description for what units this uses.
```
Comment (nathan-moore) [inline include/pros/rotation.h:105]: FYI: another TODO

File: include/pros/rotation.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+// NOTE: Descriptions of what each of these methods do may be inaccurate,
+// as it's just a speculative guess as to what they actually do.
+
+// TODO: Make revisions when proper documentation is given.
+
+/**
+ * Resets Rotational Sensor 
+ *
+ * Resets rotation presumably to 0.
```
Comment (nathan-moore) [inline include/pros/rotation.h:40]: Double check to remove the presumably?

File: include/pros/rotation.h [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+extern "C" {
+namespace pros {
+namespace c {
+#endif
+
+/**
+ * Resets Rotational Sensor 
+ *
+ * Resets rotation sensor by multiplying it by -1 only if the direction
+ * was recently reversed.
```
Comment (nathan-moore) [inline include/pros/rotation.h:35]: nit: this can be cleaner, and can you re-write without using "it"?

File: include/pros/rotation.h [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+ * reached:
+ * ENXIO - The given value is not within the range of V5 ports (1-21).
+ * ENODEV - The port cannot be configured as an Rotation Sensor
+ *
+ * \param  port
+ * 				 The V5 Rotation Sensor port number from 1-21
+ * \param  value
+ * 				 Determines if the direction of the rotational sensor is reversed or not.
+ * 
+ * \return 1 if operation succeeded or PROS_ERR_F if the operation failed, setting
```
Comment (nathan-moore) [inline include/pros/rotation.h:126]:
```suggestion
 * \return 1 if operation succeeded or PROS_ERR if the operation failed, setting
```

File: include/pros/rotation.hpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+class Rotation {
+	const std::uint8_t _port;
+
+	public:
+	Rotation(const std::uint8_t port) : _port(port){};
+
+	/**
+	 * Resets Rotational Sensor
+	 *
+	 * Resets rotation presumably to 0.
```
Comment (nathan-moore) [inline include/pros/rotation.hpp:35]:
```suggestion
	 * Resets rotation to 0.
```
? (Also shouldn't be different from the c comment)

File: include/pros/rotation.hpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	 *
+	 * \param position
+	 * 		  The position in terms of ticks
+	 * \return 1 if the operation was successful or PROS_ERR if the operation
+	 * failed, setting errno.
+	 */
+	virtual std::int32_t set_position(std::uint32_t position);
+
+	/**
+	 * Get the Rotation sensor's current rotational position.
```
Comment (nathan-moore) [inline include/pros/rotation.hpp:63]: nit: should be the same as the c api's comment

File: include/pros/rotation.hpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	 * ENXIO - The given value is not within the range of V5 ports (1-21).
+	 * ENODEV - The port cannot be configured as an Rotation Sensor
+	 *
+	 * \return The position value or PROS_ERR if the operation failed, setting
+	 * errno.
+	 */
+	virtual std::int32_t get_position();
+
+	/**
+	 * Get the Rotation sensor's current rotational velocity.
```
Comment (nathan-moore) [inline include/pros/rotation.hpp:76]: It would be nice if we can get units on this. Is it degree's per second?

**Review by nathan-moore — DISMISSED** [review]

File: include/pros/rotation.h [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+extern "C" {
+namespace pros {
+namespace c {
+#endif
+
+/**
+ * Resets Rotational Sensor 
+ *
+ * Resets rotation sensor by multiplying the rotation reading by 
+ * -1 only if the direction was recently reversed.
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:35]: at a glance i'm not really sure what this means

File: include/pros/rotation.h [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+ * \param  port
+ * 				 The V5 Rotation Sensor port number from 1-21
+ *
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t rotation_reverse(uint8_t port);
+
+/**
+ * Gets if the rotational sensor's positive/negative direction is reversed or not.
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:165]:
```suggestion
 * Check whether the Rotational sensor's direction is reversed.
```

File: include/pros/rotation.h [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+ * \param  value
+ * 				 Determines if the direction of the rotational sensor is reversed or not.
+ * 
+ * \return 1 if operation succeeded or PROS_ERR if the operation failed, setting
+ * errno.
+ */
+int32_t rotation_set_reversed(uint8_t port, bool value);
+
+/**
+ * Reverses the rotational sensor's positive counterclockwise/clockwise direction.
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:149]:
```suggestion
 * Reverse the Rotational sensor's direction.
```

if you feel the need to clarify this further, do it in another line

File: include/pros/rotation.h [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+ *
+ * \param  port
+ * 				 The V5 Rotation Sensor port number from 1-21
+ * \return The angle value or PROS_ERR_F if the operation failed, setting
+ * errno.
+ */
+int32_t rotation_get_angle(uint8_t port);
+
+/**
+ * Sets if the rotational sensor's positive/negative direction is reversed or not.
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:131]:
```suggestion
 * Set whether the Rotational sensor's direction is reversed.
```

Comment (HotelCalifornia) [reply include/pros/rotation.h:131]: is it "Rotation sensor," "rotational sensor," "Rotational sensor," "Rotation Sensor," or some other variant? pick one that is consistent with the rest of our headers and also make it consistent within this file.

File: include/pros/rotation.hpp [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+ */
+#ifndef _PROS_ROTATION_HPP_
+#define _PROS_ROTATION_HPP_
+
+#include <cstdint>
+
+#include "pros/rotation.h"
+
+namespace pros {
+class Rotation {
```
Comment (HotelCalifornia) [inline include/pros/rotation.hpp:26]: same documentation concerns as in the C header apply here

File: include/pros/rotation.h (hunk 8 above)
Comment (nathan-moore) [reply include/pros/rotation.h:35]: We were talking about this a bit on discord. From what will was saying was that it was cursed and we might need a custom implementation for now.

Comment (WillXuCodes) [reply include/pros/rotation.h:35]: (Reflected from Discord) to make up for the cursed reset function, we also now have rotation_reset_position which resets the position to 0.

File: include/pros/rotation.h [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+#ifdef __cplusplus
+extern "C" {
+namespace pros {
+namespace c {
+#endif
+
+/**
+ * Resets Rotation Sensor 
+ *
+ * Resets rotation sensor by multiplying the rotation reading by
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:34]: re: sensor naming consistency

File: include/pros/rotation.h [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+namespace pros {
+namespace c {
+#endif
+
+/**
+ * Resets Rotation Sensor
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:32]: also use the imperative mood here ("reset" as opposed to "resets")

File: include/pros/rotation.h [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+ *		  The V5 Rotation Sensor port number from 1-21
+ * \param position 
+ * 		  The position in terms of ticks
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t rotation_set_position(uint8_t port, uint32_t position);
+
+/**
+ * Resets the rotation sensor position to 0.
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:68]: imperative mood. also, if you're putting a period at the end of the first line, make sure it's at least consistent within the file, and ideally we have been consistent elsewhere in the project too

File: include/pros/rotation.h [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+ *		  The V5 Rotation Sensor port number from 1-21
+
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t rotation_reset_position(uint8_t port);
+
+/**
+ * Get the Rotation sensor's current position.  Specifically, in
+ * terms of hundreths of degrees.
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:85]:
```suggestion
 * Get the Rotation sensor's current position in centidegrees
```

re: capitalization of the sensor? be consistent

**also i am not very strongly attached to using "centidegrees" so you can use "hundredths of a degree" there instead if you want

File: include/pros/rotation.h [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+ * \param  port
+ * 				 The V5 Rotation Sensor port number from 1-21
+ * \return The position value or PROS_ERR_F if the operation failed, setting
+ * errno.
+ */
+int32_t rotation_get_position(uint8_t port);
+
+/**
+ * Get the Rotation sensor's current velocity in hundreths of
+ * degrees per second.
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:101]: could go with centidegrees here to make it more concise, but if you don't, make sure to correct "hundreth" -> "hundredth"

File: include/pros/rotation.h [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+ * \param  port
+ * 				 The V5 Rotation Sensor port number from 1-21
+ * \return The velocity value or PROS_ERR_F if the operation failed, setting
+ * errno.
+ */
+int32_t rotation_get_velocity(uint8_t port);
+
+/**
+ * Get the Rotation sensor's current position in terms of an angle
+ * measured in hundreths of degrees.
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:117]: ditto here

File: include/pros/rotation.h [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+ *
+ * \param  port
+ * 				 The V5 Rotation Sensor port number from 1-21
+ * \return The angle value or PROS_ERR_F if the operation failed, setting
+ * errno.
+ */
+int32_t rotation_get_angle(uint8_t port);
+
+/**
+ * Sets if the rotation sensor's positive/negative direction is reversed or not.
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:132]:
```suggestion
 * Set the rotation sensor's direction reversed flag
```

Comment (HotelCalifornia) [reply include/pros/rotation.h:132]: if you think this needs extra clarification, again: feel free to put it in another line

**Review by HotelCalifornia — COMMENTED** [review]
there are a couple comments here that apply to the whole file, so check the file carefully when considering them

**Comment by WillXuCodes** [comment]
Bump on review please

File: include/pros/rotation.h [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+ * reached:
+ * ENXIO - The given value is not within the range of V5 ports (1-21).
+ * ENODEV - The port cannot be configured as an rotation sensor
+ *
+ * \param port
+ *        The V5 rotation sensor port number from 1-21
+ * \return 1 if the operation was successful or PROS_ERR if the operation
+ * failed, setting errno.
+ */
+int32_t rotation_reset(uint8_t port);
```
Comment (nathan-moore) [inline include/pros/rotation.h:48]: Should we even expose this? What value does it offer?

**Review by nathan-moore — DISMISSED** [review]

File: include/pros/rotation.h (hunk 20 above)
Comment (WillXuCodes) [reply include/pros/rotation.h:48]: Maybe we should make it `rotation_flip` where it calls reset and then flips the values (with the 180 deg subtraction to angle) and sets the negative ? If not I'll just take it out, but yes I agree it doesn't add that much value other than for people who want to flip the angle or positive/negative value of the position.

Comment (HotelCalifornia) [reply include/pros/rotation.h:48]: where are you even getting this description from? as far as i can tell this is not what the SDK documentation says this function does

File: include/pros/rotation.h [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+namespace pros {
+namespace c {
+#endif
+
+/**
+ * Reset rotation sensor
```
Comment (HotelCalifornia) [inline include/pros/rotation.h:32]: i changed my mind from the last review i left---capitalize Rotation Sensor throughout this file to be consistent with how we've done other sensors

Comment (WillXuCodes) [reply include/pros/rotation.h]: Done

**Review by HotelCalifornia — APPROVED** [review]
lgtm

**Pull request merged**